### PR TITLE
Mark classes as @final and deprecate extending them

### PR DESCRIPTION
**Pull request by malarzm**
Precedes #1984

### CODE REVIEW

**Review by greg0ire — APPROVED** [review]

File: lib/Doctrine/ODM/MongoDB/DocumentNotFoundException.php [hunk 1]
```diff
@@ -19,14 +19,25 @@
...
  *
+ * @final
  * @since       1.0
  */
 class DocumentNotFoundException extends MongoDBException
 {
+    public function __construct($message = "", $code = 0, Throwable $previous = null)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/DocumentNotFoundException.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/DocumentNotFoundEventArgs.php [hunk 2]
```diff
@@ -49,6 +50,9 @@ class DocumentNotFoundEventArgs extends LifecycleEventArgs
      */
     public function __construct($document, DocumentManager $dm, $identifier)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/DocumentNotFoundEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/LoadClassMetadataEventArgs.php [hunk 3]
```diff
@@ -20,14 +20,25 @@
...
  *
+ * @final
  * @since 1.0
  */
 class LoadClassMetadataEventArgs extends BaseLoadClassMetadataEventArgs
 {
+    public function __construct(ClassMetadata $classMetadata, ObjectManager $objectManager)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/LoadClassMetadataEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php [hunk 4]
```diff
@@ -29,6 +31,9 @@ class OnClassMetadataNotFoundEventArgs extends ManagerEventArgs
      */
     public function __construct($className, DocumentManager $dm)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/OnClassMetadataNotFoundEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/OnClearEventArgs.php [hunk 5]
```diff
@@ -24,10 +24,19 @@
...
  *
+ * @final
  * @since 1.0
  */
 class OnClearEventArgs extends BaseOnClearEventArgs
 {
+    public function __construct($objectManager, $entityClass = null)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/OnClearEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/OnFlushEventArgs.php [hunk 6]
```diff
@@ -19,11 +19,21 @@
...
  *
+ * @final
  * @since 1.0
  */
 class OnFlushEventArgs extends ManagerEventArgs
 {
+    public function __construct(ObjectManager $objectManager)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/OnFlushEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/PostCollectionLoadEventArgs.php [hunk 7]
```diff
@@ -23,6 +24,9 @@ class PostCollectionLoadEventArgs extends ManagerEventArgs
      */
     public function __construct(PersistentCollectionInterface $collection, DocumentManager $dm)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/PostCollectionLoadEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/PostFlushEventArgs.php [hunk 8]
```diff
@@ -19,11 +19,21 @@
...
  *
+ * @final
  * @since 1.0
  */
 class PostFlushEventArgs extends ManagerEventArgs
 {
+    public function __construct(ObjectManager $objectManager)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/PostFlushEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/PreFlushEventArgs.php [hunk 9]
```diff
@@ -19,11 +19,21 @@
...
  *
+ * @final
  * @since 1.0
  */
 class PreFlushEventArgs extends ManagerEventArgs
 {
+    public function __construct(ObjectManager $objectManager)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/PreFlushEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/PreLoadEventArgs.php [hunk 10]
```diff
@@ -42,6 +43,9 @@ class PreLoadEventArgs extends LifecycleEventArgs
      */
     public function __construct($document, DocumentManager $dm, array &$data)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/PreLoadEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Event/PreUpdateEventArgs.php [hunk 11]
```diff
@@ -42,6 +43,9 @@ class PreUpdateEventArgs extends LifecycleEventArgs
      */
     public function __construct($document, DocumentManager $dm, array $changeSet)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Event/PreUpdateEventArgs.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Hydrator/HydratorException.php [hunk 12]
```diff
@@ -20,14 +20,24 @@
...
  *
+ * @final
  * @since  1.0
  */
 class HydratorException extends MongoDBException
 {
+    public function __construct($message = "", $code = 0, Throwable $previous = null)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Hydrator/HydratorException.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php [hunk 13]
```diff
@@ -97,6 +98,9 @@ class HydratorFactory
      */
     public function __construct(DocumentManager $dm, EventManager $evm, $hydratorDir, $hydratorNs, $autoGenerate)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Hydrator/HydratorFactory.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Id/AlnumGenerator.php [hunk 14]
```diff
@@ -46,6 +47,13 @@ class AlnumGenerator extends IncrementGenerator
 
     protected $awkwardSafeChars = '0123456789BCDFGHJKLMNPQRSTVWXZbcdfghjklmnpqrstvwxz';
 
+    public function __construct()
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Id/AlnumGenerator.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Id/AutoGenerator.php [hunk 15]
```diff
@@ -24,10 +24,18 @@
...
  *
+ * @final
  * @since       1.0
  */
 class AutoGenerator extends AbstractIdGenerator
 {
+    public function __construct()
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Id/AutoGenerator.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Id/IncrementGenerator.php [hunk 16]
```diff
@@ -40,6 +41,13 @@ class IncrementGenerator extends AbstractIdGenerator
     protected $key = null;
     protected $startingId = 1;
 
+    public function __construct()
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Id/IncrementGenerator.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Id/UuidGenerator.php [hunk 17]
```diff
@@ -35,6 +36,13 @@ class UuidGenerator extends AbstractIdGenerator
      */
     protected $salt = null;
 
+    public function __construct()
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Id/UuidGenerator.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/LockException.php [hunk 18]
```diff
@@ -30,6 +31,9 @@ class LockException extends MongoDBException
 
     public function __construct($msg, $document = null)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/LockException.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataFactory.php [hunk 19]
```diff
@@ -56,6 +57,13 @@ class ClassMetadataFactory extends AbstractClassMetadataFactory
     /** @var \Doctrine\Common\EventManager The event manager instance */
     private $evm;
 
+    public function __construct()
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/ClassMetadataFactory.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Mapping/MappingException.php [hunk 20]
```diff
@@ -20,14 +20,24 @@
...
  *
+ * @final
  * @since       1.0
  */
 class MappingException extends BaseMappingException
 {
+    public function __construct($message = "", $code = 0, Throwable $previous = null)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Mapping/MappingException.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/PersistentCollection.php [hunk 21]
```diff
@@ -41,6 +42,9 @@ class PersistentCollection implements PersistentCollectionInterface
      */
     public function __construct(BaseCollection $coll, DocumentManager $dm, UnitOfWork $uow)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/PersistentCollection.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/PersistentCollection/PersistentCollectionException.php [hunk 22]
```diff
@@ -20,14 +20,24 @@
...
  *
+ * @final
  * @since 1.1
  */
 class PersistentCollectionException extends MongoDBException
 {
+    public function __construct($message = "", $code = 0, Throwable $previous = null)
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/PersistentCollection/PersistentCollectionException.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Persisters/CollectionPersister.php [hunk 23]
```diff
@@ -78,6 +79,9 @@ class CollectionPersister
      */
     public function __construct(DocumentManager $dm, PersistenceBuilder $pb, UnitOfWork $uow)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Persisters/CollectionPersister.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php [hunk 24]
```diff
@@ -136,6 +137,9 @@ public function __construct(
         ClassMetadata $class,
         CriteriaMerger $cm = null
     ) {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Persisters/DocumentPersister.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Persisters/PersistenceBuilder.php [hunk 25]
```diff
@@ -56,6 +57,9 @@ class PersistenceBuilder
      */
     public function __construct(DocumentManager $dm, UnitOfWork $uow)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Persisters/PersistenceBuilder.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Query/CriteriaMerger.php [hunk 26]
```diff
@@ -25,9 +25,18 @@
...
  * where criteria might be inadvertently overridden with array_merge().
+ *
+ * @final
  */
 class CriteriaMerger
 {
+    public function __construct()
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Query/CriteriaMerger.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Query/FilterCollection.php [hunk 27]
```diff
@@ -56,6 +58,9 @@ class FilterCollection
      */
     public function __construct(DocumentManager $dm, CriteriaMerger $cm = null)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Query/FilterCollection.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Query/Query.php [hunk 28]
```diff
@@ -107,6 +108,10 @@ class Query extends \Doctrine\MongoDB\Query\Query
      */
     public function __construct(DocumentManager $dm, ClassMetadata $class, Collection $collection, array $query = array(), array $options = array(), $hydrate = true, $refresh = false, array $primers = array(), $requireIndexes = null, $readOnly = false)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Query/Query.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Query/QueryExpressionVisitor.php [hunk 29]
```diff
@@ -71,6 +72,9 @@ class QueryExpressionVisitor extends ExpressionVisitor
      */
     public function __construct(Builder $builder)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Query/QueryExpressionVisitor.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php [hunk 30]
```diff
@@ -69,6 +70,10 @@ class ReferencePrimer
      */
     public function __construct(DocumentManager $dm, UnitOfWork $uow)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Query/ReferencePrimer.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/SchemaManager.php [hunk 31]
```diff
@@ -55,6 +58,9 @@ class SchemaManager
      */
     public function __construct(DocumentManager $dm, ClassMetadataFactory $cmf)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/SchemaManager.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/UnitOfWork.php [hunk 32]
```diff
@@ -276,6 +277,9 @@ class UnitOfWork implements PropertyChangedListener
      */
     public function __construct(DocumentManager $dm, EventManager $evm, HydratorFactory $hydratorFactory)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/UnitOfWork.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Utility/CollectionHelper.php [hunk 33]
```diff
@@ -24,12 +24,20 @@
...
  * @internal
  */
 class CollectionHelper
 {
     const DEFAULT_STRATEGY = ClassMetadata::STORAGE_STRATEGY_PUSH_ALL;
 
+    public function __construct()
+    {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Utility/CollectionHelper.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

File: lib/Doctrine/ODM/MongoDB/Utility/LifecycleEventManager.php [hunk 34]
```diff
@@ -58,6 +59,9 @@ class LifecycleEventManager
      */
     public function __construct(DocumentManager $dm, UnitOfWork $uow, EventManager $evm)
     {
+        if (self::class !== static::class) {
+            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in 2.0.', static::class, self::class), E_USER_DEPRECATED);
```
Comment (alcaeus) [inline lib/Doctrine/ODM/MongoDB/Utility/LifecycleEventManager.php]:
```suggestion
            @trigger_error(sprintf('The class "%s" extends "%s" which will be final in MongoDB ODM 2.0.', static::class, self::class), E_USER_DEPRECATED);
```

**Review by alcaeus — CHANGES_REQUESTED** [review]
Missed a couple of package names in the deprecation messages 😅

**Comment by malarzm** [comment]
@alcaeus I think it'd be the first place where we add "MongoDB ODM" to the deprecation message, why now?

**Comment by malarzm** [comment]
Ah nvm, just found #2010. Maybe I should go newest -> oldest when reading notifications :D

**Comment by alcaeus** [comment]
Yeah, that was added afterwards when I gave it some more thought. I’ll fix those messages today or tomorrow, just thought that we should start somewhere 👍

**Review by alcaeus — APPROVED** [review]
Looks good, thanks @malarzm!

**Pull request merged**